prototype2/common/EV42Serializer.cpp

```diff
@@ -0,0 +1,106 @@
+/** Copyright (C) 2016, 2017 European Spallation Source ERIC */
+
+#include <common/EV42Serializer.h>
+#include "ev42_events_generated.h"
+#include <libs/include/gccintel.h>
+
+#include <common/Trace.h>
+//#undef TRC_LEVEL
+//#define TRC_LEVEL TRC_L_DEB
+
+static constexpr size_t TimeSize = sizeof(uint32_t);
+static constexpr size_t PixelSize = sizeof(uint32_t);
+
+// These must be non-0 because of Flatbuffers being stupid
+// If they are initially set to 0, they will not be mutable
+static constexpr uint64_t FBMutablePlaceholder = 1;
+
+static_assert(FLATBUFFERS_LITTLEENDIAN,
+              "Flatbuffers only tested on little endian systems");
+
+EV42Serializer::EV42Serializer(size_t max_array_length, std::string source_name)
+    : MaxEvents(max_array_length), Builder_(MaxEvents * 8 + 256) {
+
+  auto sourceName = Builder_.CreateString(source_name);
+  auto timeoff = Builder_.CreateUninitializedVector(MaxEvents, TimeSize, &TimePtr);
+  auto pixeloff =
+      Builder_.CreateUninitializedVector(MaxEvents, PixelSize, &PixelPtr);
+
+  auto evMsgHeader = CreateEventMessage(Builder_, sourceName, FBMutablePlaceholder,
+                                        FBMutablePlaceholder, timeoff, pixeloff);
+  FinishEventMessageBuffer(Builder_, evMsgHeader);
+
+  Buffer_.address = Builder_.GetBufferPointer();
+  Buffer_.size = Builder_.GetSize();
+  assert(Buffer_);
+
+  EventMessage_ = const_cast<EventMessage *>(GetEventMessage(Buffer_.address));
+  TimeLengthPtr =
+      reinterpret_cast<flatbuffers::uoffset_t *>(
+          const_cast<std::uint8_t *>(EventMessage_->time_of_flight()->Data())) - 1;
+  PixelLengthPtr =
+      reinterpret_cast<flatbuffers::uoffset_t *>(
+          const_cast<std::uint8_t *>(EventMessage_->detector_id()->Data())) - 1;
+
+  EventMessage_->mutate_message_id(0);
+  EventMessage_->mutate_pulse_time(0);
+}
+
+void EV42Serializer::producerCallback(ProducerCallback callback) {
```

Comment:
I think that renaming the function to (e.g.) `setProducerCallback()` would be slight improvement.